File: features/admin/config/permissions.ts
/**
 * ADM-110 — FE mirror of backend authz.AllPermissionCodes.
 * Backend remains authority; FE only hides/disables existing controls.
 */

/** Canonical admin + shared permission codes (launch registry). */
export const ALL_PERMISSION_CODES = [
  "admin.ping",
  "admin.dashboard.read",
  "merchants.read",
  "merchants.write",
  "buyers.read",
  "orders.read",
  "payments.read",
  "kyc.review",
  "withdrawals.review",
  "impersonation.start",
  "impersonation.support_write",
  "provider_callbacks.replay",
  "seller_webhook_deliveries.retry",
  "webhooks.read",
  "roles.read",
  "roles.write",
  "roles.assign",
  "users.read",
  "fulfillment.force",
  "fulfillment.read",
  "inventory.reveal",
  "inventory.read",
  "reviews.read",
  "reviews.moderate",
  "campaigns.publish",
  "platform.emergency",
  "platform.fees.preview",
  "audit.read",
  "seller.store.read",
  "seller.store.write",
  "buyer.purchases.read",
  "invitations.staff",
  "invitations.merchant",
] as const;

export type PermissionCode = (typeof ALL_PERMISSION_CODES)[number];

const KNOWN = new Set<string>(ALL_PERMISSION_CODES);

/** Superuser wildcard from session claims (mock / SUPER_ADMIN bootstrap). */
export const PERMISSION_WILDCARD = "*";

export function isKnownPermissionCode(code: string): boolean {
  return KNOWN.has(code);
}

/**
 * Fail-closed permission check against session claim list.
 * Unknown codes never grant access (even with wildcard).
 */
export function claimsHavePermission(
  permissions: readonly string[] | null | undefined,
  code: string | null | undefined,
): boolean {
  if (!code) return false;
  if (!isKnownPermissionCode(code)) return false;
  if (!permissions || permissions.length === 0) return false;
  if (permissions.includes(PERMISSION_WILDCARD)) return true;
  return permissions.includes(code);
}

/**
 * Authenticated-admin surface gate (profile / subject-owned me routes).
 * No dedicated page permission code in BE registry.
 */
export function claimsAreAuthenticatedAdmin(claims: {
  subjectId?: string | null;
  surface?: string | null;
} | null | undefined): boolean {
  return Boolean(claims?.subjectId && claims.surface === "admin");
}

/** Action-level permission keys used by existing admin controls. */
export const ADMIN_ACTION_PERMISSIONS = {
  merchantsWrite: "merchants.write",
  rolesWrite: "roles.write",
  rolesAssign: "roles.assign",
  staffInvite: "roles.assign",
  auditExport: "audit.read",
  kycReview: "kyc.review",
  withdrawalsReview: "withdrawals.review",
  reviewsModerate: "reviews.moderate",
  inventoryReveal: "inventory.reveal",
  fulfillmentForce: "fulfillment.force",
  providerCallbacksReplay: "provider_callbacks.replay",
  sellerWebhookRetry: "seller_webhook_deliveries.retry",
  impersonationStart: "impersonation.start",
  platformEmergency: "platform.emergency",
  platformFeesPreview: "platform.fees.preview",
  campaignsPublish: "campaigns.publish",
} as const satisfies Record<string, PermissionCode>;

export type AdminActionPermissionKey = keyof typeof ADMIN_ACTION_PERMISSIONS;

File: backend/internal/domain/authz/permissions.go
// Package authz owns RBAC permission codes, system roles, and tenant policy (BE-130).
package authz

// Stable permission codes (registry source of truth for seeds + matrix).
const (
	PermAdminPing                    = "admin.ping"
	PermAdminDashboardRead           = "admin.dashboard.read"
	PermMerchantsRead                = "merchants.read"
	PermMerchantsWrite               = "merchants.write"
	PermBuyersRead                   = "buyers.read"
	PermOrdersRead                   = "orders.read"
	PermPaymentsRead                 = "payments.read"
	PermKYCReview                    = "kyc.review"
	PermWithdrawalsReview            = "withdrawals.review"
	PermImpersonationStart           = "impersonation.start"
	PermImpersonationSupportWrite    = "impersonation.support_write"
	PermProviderCallbacksReplay      = "provider_callbacks.replay"
	PermSellerWebhookDeliveriesRetry = "seller_webhook_deliveries.retry"
	PermWebhooksRead                 = "webhooks.read"
	PermRolesRead                    = "roles.read"
	PermRolesWrite                   = "roles.write"
	PermRolesAssign                  = "roles.assign"
	PermUsersRead                    = "users.read"
	PermFulfillmentForce             = "fulfillment.force"
	PermFulfillmentRead              = "fulfillment.read"
	PermInventoryReveal              = "inventory.reveal"
	PermInventoryRead                = "inventory.read"
	PermReviewsRead                  = "reviews.read"
	PermReviewsModerate              = "reviews.moderate"
	PermCampaignsPublish             = "campaigns.publish"
	PermPlatformEmergency            = "platform.emergency"
	PermPlatformFeesPreview          = "platform.fees.preview"
	PermAuditRead                    = "audit.read"
	PermSellerStoreRead              = "seller.store.read"
	PermSellerStoreWrite             = "seller.store.write"
	PermBuyerPurchasesRead           = "buyer.purchases.read"
	PermInvitationsStaff             = "invitations.staff"
	PermInvitationsMerchant          = "invitations.merchant"
)

// System role codes (immutable; is_system=true in DB).
const (
	RoleSuperAdmin   = "SUPER_ADMIN"
	RoleAdminSupport = "ADMIN_SUPPORT"
	RoleAdminFinance = "ADMIN_FINANCE"
	RoleSellerOwner  = "SELLER_OWNER"
	RoleBuyer        = "BUYER"
)

// SystemRoleIDs are stable primary keys used by migration seed.
const (
	RoleIDSuperAdmin   = "role_super_admin"
	RoleIDAdminSupport = "role_admin_support"
	RoleIDAdminFinance = "role_admin_finance"
	RoleIDSellerOwner  = "role_seller_owner"
	RoleIDBuyer        = "role_buyer"
)

// MerchantMemberRole is membership within a merchant tenant.
type MerchantMemberRole string

const (
	MemberOwner MerchantMemberRole = "OWNER"
	MemberStaff MerchantMemberRole = "STAFF"
)

// MerchantMemberStatus is membership lifecycle.
type MerchantMemberStatus string

const (
	MemberActive    MerchantMemberStatus = "ACTIVE"
	MemberInvited   MerchantMemberStatus = "INVITED"
	MemberSuspended MerchantMemberStatus = "SUSPENDED"
	MemberRemoved   MerchantMemberStatus = "REMOVED"
)

// MerchantStatus is merchant lifecycle.
type MerchantStatus string

const (
	MerchantActive    MerchantStatus = "ACTIVE"
	MerchantSuspended MerchantStatus = "SUSPENDED"
	MerchantClosed    MerchantStatus = "CLOSED"
)

// StoreStatus is store lifecycle.
type StoreStatus string

const (
	StoreActive    StoreStatus = "ACTIVE"
	StoreSuspended StoreStatus = "SUSPENDED"
	StoreArchived  StoreStatus = "ARCHIVED"
)

// AllPermissionCodes returns the launch permission registry (order stable for tests).
func AllPermissionCodes() []string {
	return []string{
		PermAdminPing,
		PermAdminDashboardRead,
		PermMerchantsRead,
		PermMerchantsWrite,
		PermBuyersRead,
		PermOrdersRead,
		PermPaymentsRead,
		PermKYCReview,
		PermWithdrawalsReview,
		PermImpersonationStart,
		PermImpersonationSupportWrite,
		PermProviderCallbacksReplay,
		PermSellerWebhookDeliveriesRetry,
		PermWebhooksRead,
		PermRolesRead,
		PermRolesWrite,
		PermRolesAssign,
		PermUsersRead,
		PermFulfillmentForce,
		PermFulfillmentRead,
		PermInventoryReveal,
		PermInventoryRead,
		PermReviewsRead,
		PermReviewsModerate,
		PermCampaignsPublish,
		PermPlatformEmergency,
		PermPlatformFeesPreview,
		PermAuditRead,
		PermSellerStoreRead,
		PermSellerStoreWrite,
		PermBuyerPurchasesRead,
		PermInvitationsStaff,
		PermInvitationsMerchant,
	}
}

// SystemRoleCodes returns immutable system role codes.
func SystemRoleCodes() []string {
	return []string{
		RoleSuperAdmin,
		RoleAdminSupport,
		RoleAdminFinance,
		RoleSellerOwner,
		RoleBuyer,
	}
}

// IsSystemRole reports whether code is a seeded system role.
func IsSystemRole(code string) bool {
	switch code {
	case RoleSuperAdmin, RoleAdminSupport, RoleAdminFinance, RoleSellerOwner, RoleBuyer:
		return true
	default:
		return false
	}
}

// HasPermission reports whether set contains code (deny by default).
func HasPermission(set map[string]struct{}, code string) bool {
	if set == nil || code == "" {
		return false
	}
	_, ok := set[code]
	return ok
}

// PermissionSet builds a set from a slice.
func PermissionSet(codes []string) map[string]struct{} {
	out := make(map[string]struct{}, len(codes))
	for _, c := range codes {
		if c != "" {
			out[c] = struct{}{}
		}
	}
	return out
}
